File: teaching/anthropic/anthropic-protocol-book/code/python/ch06/tool_registry.py
"""
Chapter 6: Tool Registry.

Protocol-level implementation of Anthropic's Tool Use feature, managing
tool definitions, deferred loading strategies, allowed caller configuration,
and providing unified API format conversion for the Messages API.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, Dict, List, Optional


@dataclass
class ToolDefinition:
    """Complete definition of a single tool.

    Mirrors the Anthropic Messages API tool object, with additional
    metadata for deferred loading, programmatic calling, and examples.
    """

    name: str
    description: str
    input_schema: Dict[str, Any]
    handler: Callable[..., Any]
    defer_loading: bool = False
    allowed_callers: Optional[List[str]] = None
    input_examples: Optional[List[Dict[str, Any]]] = None
    strict: bool = False

    def to_api_format(self) -> dict:
        """Convert to Anthropic API-compatible tool dictionary."""
        tool: Dict[str, Any] = {
            "name": self.name,
            "description": self.description,
            "input_schema": self.input_schema,
        }
        if self.defer_loading:
            tool["defer_loading"] = True
        if self.allowed_callers:
            tool["allowed_callers"] = self.allowed_callers
        if self.input_examples:
            tool["input_examples"] = self.input_examples
        if self.strict:
            tool["strict"] = True
        return tool


class ToolRegistry:
    """Registry for managing tool definitions, execution, and API formatting.

    Acts as the central authority for all tools available to Claude.
    Supports standard tool use, strict tool use, deferred loading
    (for Tool Search Tool), and programmatic tool calling.

    Usage::

        registry = ToolRegistry()
        registry.register(
            name="get_weather",
            description="Get current weather for a location",
            input_schema={
                "type": "object",
                "properties": {
                    "location": {"type": "string", "description": "City name"}
                },
                "required": ["location"],
            },
            handler=lambda location: f"Weather for {location}: 72F",
            defer_loading=False,
        )

        # Generate API-compatible tools list
        tools = registry.to_api_format()

        # Execute a tool
        result = registry.execute("get_weather", {"location": "SF"})
    """

    def __init__(self) -> None:
        self._tools: Dict[str, ToolDefinition] = {}
        self._search_tool_enabled: bool = False
        self._code_execution_enabled: bool = False

    # ------------------------------------------------------------------
    # Registration
    # ------------------------------------------------------------------

    def register(
        self,
        name: str,
        description: str,
        input_schema: Dict[str, Any],
        handler: Callable[..., Any],
        defer_loading: bool = False,
        allowed_callers: Optional[List[str]] = None,
        input_examples: Optional[List[Dict[str, Any]]] = None,
        strict: bool = False,
    ) -> None:
        """Register a tool.

        Args:
            name: Unique tool name. Use ``domain_action`` pattern.
            description: Detailed description including when to use and return format.
            input_schema: JSON Schema ``{"type": "object", "properties": {...}}``.
            handler: Function that receives ``**kwargs`` and returns the result.
            defer_loading: If True, tool is discoverable via Tool Search only.
            allowed_callers: ``["direct"]`` and/or ``["code_execution_20260120"]``.
            input_examples: Array of example input objects for Tool Use Examples.
            strict: If True, enables grammar-constrained sampling.

        Raises:
            ValueError: If tool name is already registered, handler is not callable,
                        or input_schema lacks proper structure.
        """
        if name in self._tools:
            raise ValueError(f"Tool '{name}' is already registered")
        if not callable(handler):
            raise ValueError(
                f"Handler for tool '{name}' must be callable"
            )
        if not isinstance(input_schema, dict) or input_schema.get("type") != "object":
            raise ValueError(
                "input_schema must be a JSON Schema object with type='object'"
            )

        self._tools[name] = ToolDefinition(
            name=name,
            description=description,
            input_schema=input_schema,
            handler=handler,
            defer_loading=defer_loading,
            allowed_callers=allowed_callers,
            input_examples=input_examples,
            strict=strict,
        )

    def unregister(self, name: str) -> None:
        """Remove a previously registered tool (no-op if not found)."""
        self._tools.pop(name, None)

    # ------------------------------------------------------------------
    # Queries
    # ------------------------------------------------------------------

    def get(self, name: str) -> Optional[ToolDefinition]:
        """Retrieve a tool definition by name, or None."""
        return self._tools.get(name)

    def list_names(self) -> List[str]:
        """Return all registered tool names in insertion order."""
        return list(self._tools.keys())

    def get_deferred_tools(self) -> List[ToolDefinition]:
        """Return tools marked with ``defer_loading: True``."""
        return [t for t in self._tools.values() if t.defer_loading]

    def get_immediate_tools(self) -> List[ToolDefinition]:
        """Return tools that are NOT deferred (loaded into context immediately)."""
        return [t for t in self._tools.values() if not t.defer_loading]

    def get_programmatic_tools(self) -> List[ToolDefinition]:
        """Return tools callable from the code execution environment."""
        return [
            t
            for t in self._tools.values()
            if t.allowed_callers and "code_execution_20260120" in t.allowed_callers
        ]

    def get_strict_tools(self) -> List[ToolDefinition]:
        """Return tools with strict mode enabled."""
        return [t for t in self._tools.values() if t.strict]

    def count(self) -> int:
        """Total number of registered tools."""
        return len(self._tools)

    # ------------------------------------------------------------------
    # API format
    # ------------------------------------------------------------------

    def to_api_format(self) -> List[Dict[str, Any]]:
        """Generate the tools array for the Anthropic Messages API.

        Returns all tools (including deferred ones) because the API
        needs their full definitions to expand ``tool_reference`` blocks.
        """
        return [tool.to_api_format() for tool in self._tools.values()]

    def to_search_index_format(self) -> List[Dict[str, Any]]:
        """Generate tool metadata suitable for building a search index.

        Only includes deferred tools, as non-deferred tools are already
        in the context window.
        """
        result: List[Dict[str, Any]] = []
        for tool in self._tools.values():
            if tool.defer_loading:
                result.append(
                    {
                        "name": tool.name,
                        "description": tool.description,
                        "input_schema": tool.input_schema,
                    }
                )
        return result

    # ------------------------------------------------------------------
    # Execution
    # ------------------------------------------------------------------

    def execute(self, tool_name: str, input_data: Dict[str, Any]) -> Dict[str, Any]:
        """Execute a tool's handler function.

        Args:
            tool_name: Registered tool name.
            input_data: Keyword arguments to pass to the handler.

        Returns:
            ``{"success": True, "result": ...}`` or
            ``{"success": False, "error": "..."}``.

        Raises:
            KeyError: If tool_name is not registered.
        """
        tool = self._tools.get(tool_name)
        if tool is None:
            raise KeyError(f"Tool '{tool_name}' not found in registry")
        try:
            result = tool.handler(**input_data)
            return {"success": True, "result": result}
        except Exception as exc:
            return {"success": False, "error": str(exc)}

    def execute_many(
        self, calls: List[Dict[str, Any]]
    ) -> List[Dict[str, Any]]:
        """Execute multiple tool calls (for parallel tool use support).

        Args:
            calls: List of ``{"tool_name": "...", "input": {...}}``.

        Returns:
            Results in the same order as the input calls.
        """
        return [self.execute(c["tool_name"], c["input"]) for c in calls]

    # ------------------------------------------------------------------
    # Feature toggles
    # ------------------------------------------------------------------

    def enable_tool_search(self) -> None:
        """Mark that Tool Search Tool is enabled."""
        self._search_tool_enabled = True

    def enable_code_execution(self) -> None:
        """Mark that Code Execution Tool is enabled."""
        self._code_execution_enabled = True

    @property
    def is_search_enabled(self) -> bool:
        return self._search_tool_enabled

    @property
    def is_code_execution_enabled(self) -> bool:
        return self._code_execution_enabled

File: teaching/anthropic/anthropic-protocol-book/code/node/ch06/ToolRegistry.ts
/**
 * Chapter 6: Tool Registry (TypeScript).
 *
 * Protocol-level implementation of Anthropic's Tool Use feature.
 * Manages tool definitions, deferred loading, allowed callers,
 * and API format conversion for the Messages API.
 */

export interface JsonSchema {
  type: "object";
  properties: Record<string, Record<string, unknown>>;
  required?: string[];
  additionalProperties?: boolean;
}

export interface ToolDefinition {
  name: string;
  description: string;
  input_schema: JsonSchema;
  handler: (...args: any[]) => any;
  defer_loading?: boolean;
  allowed_callers?: string[];
  input_examples?: Record<string, unknown>[];
  strict?: boolean;
}

export interface ToolCall {
  tool_name: string;
  input: Record<string, unknown>;
}

export interface ToolResult {
  success: boolean;
  result?: unknown;
  error?: string;
}

const REQUIRED_CODE_EXEC_CALLER = "code_execution_20260120";

/**
 * Central registry for tool definitions, execution, and API formatting.
 *
 * Usage:
 * ```typescript
 * const registry = new ToolRegistry();
 * registry.register("get_weather", "Get weather for a location",
 *   { type: "object", properties: { location: { type: "string" } }, required: ["location"] },
 *   (input) => `Weather: 72F`
 * );
 * const tools = registry.toApiFormat();
 * ```
 */
export class ToolRegistry {
  private tools: Map<string, ToolDefinition> = new Map();
  private searchEnabled: boolean = false;
  private codeExecutionEnabled: boolean = false;

  // ------------------------------------------------------------------
  // Registration
  // ------------------------------------------------------------------

  register(
    name: string,
    description: string,
    input_schema: JsonSchema,
    handler: (...args: any[]) => any,
    options: {
      defer_loading?: boolean;
      allowed_callers?: string[];
      input_examples?: Record<string, unknown>[];
      strict?: boolean;
    } = {},
  ): void {
    if (this.tools.has(name)) {
      throw new Error(`Tool '${name}' is already registered`);
    }
    if (typeof handler !== "function") {
      throw new Error(`Handler for tool '${name}' must be a function`);
    }
    if (!input_schema || input_schema.type !== "object") {
      throw new Error(
        "input_schema must be a JSON Schema object with type='object'",
      );
    }

    this.tools.set(name, {
      name,
      description,
      input_schema,
      handler,
      defer_loading: options.defer_loading ?? false,
      allowed_callers: options.allowed_callers,
      input_examples: options.input_examples,
      strict: options.strict ?? false,
    });
  }

  unregister(name: string): void {
    this.tools.delete(name);
  }

  // ------------------------------------------------------------------
  // Queries
  // ------------------------------------------------------------------

  get(name: string): ToolDefinition | undefined {
    return this.tools.get(name);
  }

  listNames(): string[] {
    return Array.from(this.tools.keys());
  }

  getDeferredTools(): ToolDefinition[] {
    return Array.from(this.tools.values()).filter((t) => t.defer_loading);
  }

  getImmediateTools(): ToolDefinition[] {
    return Array.from(this.tools.values()).filter((t) => !t.defer_loading);
  }

  getProgrammaticTools(): ToolDefinition[] {
    return Array.from(this.tools.values()).filter(
      (t) =>
        t.allowed_callers !== undefined &&
        t.allowed_callers.includes(REQUIRED_CODE_EXEC_CALLER),
    );
  }

  getStrictTools(): ToolDefinition[] {
    return Array.from(this.tools.values()).filter((t) => t.strict);
  }

  count(): number {
    return this.tools.size;
  }

  // ------------------------------------------------------------------
  // API format
  // ------------------------------------------------------------------

  toApiFormat(): Record<string, unknown>[] {
    const result: Record<string, unknown>[] = [];
    for (const tool of this.tools.values()) {
      const entry: Record<string, unknown> = {
        name: tool.name,
        description: tool.description,
        input_schema: tool.input_schema,
      };
      if (tool.defer_loading) entry["defer_loading"] = true;
      if (tool.allowed_callers) entry["allowed_callers"] = tool.allowed_callers;
      if (tool.input_examples) entry["input_examples"] = tool.input_examples;
      if (tool.strict) entry["strict"] = true;
      result.push(entry);
    }
    return result;
  }

  toSearchIndexFormat(): Record<string, unknown>[] {
    const result: Record<string, unknown>[] = [];
    for (const tool of this.tools.values()) {
      if (tool.defer_loading) {
        result.push({
          name: tool.name,
          description: tool.description,
          input_schema: tool.input_schema,
        });
      }
    }
    return result;
  }

  // ------------------------------------------------------------------
  // Execution
  // ------------------------------------------------------------------

  execute(toolName: string, inputData: Record<string, unknown>): ToolResult {
    const tool = this.tools.get(toolName);
    if (!tool) {
      throw new Error(`Tool '${toolName}' not found in registry`);
    }
    try {
      const result = tool.handler(inputData);
      return { success: true, result };
    } catch (e: unknown) {
      return {
        success: false,
        error: e instanceof Error ? e.message : String(e),
      };
    }
  }

  executeMany(calls: ToolCall[]): ToolResult[] {
    return calls.map((call) => this.execute(call.tool_name, call.input));
  }

  // ------------------------------------------------------------------
  // Feature toggles
  // ------------------------------------------------------------------

  enableToolSearch(): void {
    this.searchEnabled = true;
  }

  enableCodeExecution(): void {
    this.codeExecutionEnabled = true;
  }

  get isSearchEnabled(): boolean {
    return this.searchEnabled;
  }

  get isCodeExecutionEnabled(): boolean {
    return this.codeExecutionEnabled;
  }
}
